### DISCUSSION TITLE
What's the right way to use `TetGen`?

### DISCUSSION THREAD

**Discussion by Xayah-Hina** [InteractiveComputerGraphics/PositionBasedDynamics#109] [General]
Hi~

I'm working on a FEM simulation on `PositionBasedDynamics` and get trapped by the right way to convert a `obj` mesh to a `tet model`.

In the example scene `DeformableSolidCollisionScene`, I learned that `.ele` file and `.node` file are required to get a tetrahedron model, they're produced by the tool [`TetGen`](http://wias-berlin.de/software/index.jsp?id=TetGen&lang=1#Documentation).

But unfortunately, I found that `TetGen` seems not to support a direct approach to convert a `.obj` file to `.ele` and `.node` [TetGen doc](http://www.tetgen.org/1.5/doc/manual/manual.pdf), it seems only support `.off` `.ply` `.stl` `.mesh`.  And there is a command `tetgen -pqa5000.0 Armadillo_UV_low.off ` at the last line of the files `armadillo_4k.ele` and `armadillo_4k.node`, which hints that a `.off` is required, but I don't know where can I get the `.off` file...

---
So my confusion is that,
- how can I use `TetGen ` correctly when working on `PositionBasedDynamics`?

more specifically,
- what is the `.off` file?
- is it simply converted from a `.obj` file or whatever else?
- what is the most recommended way to do with `TetGen` and `PositionBasedDynamics`?

Best regards~

**Comment by korzen** [comment]
Hi, you can convert obj to ply and/or stl using e.g. free MeshLab.

TetGen has great documention http://wias-berlin.de/software/tetgen/1.5/doc/manual/manual006.html#sec88

**Comment by Xayah-Hina** [comment]
@korzen Oh yes, I used `MeshLab`. my question now is: why don't the `visFile` in `xxxScene.json` directly support `ply` or `off` format? 

and btw, if I want to realize **dynamic tetrahedralization** and **fracture** like this ->  [Adaptive Tetrahedral Meshes for Brittle Fracture Simulation](https://animation.rwth-aachen.de/publication/059/), I wonder if there is any convenient approach for `PositionBasedDynamics` framework?

Sorry to bother you for my personal problems~ I'm new to physics based animation, and really appreciate your help! \~\(≥▽≤)/\~

**Comment by janbender** [comment]
> @korzen Oh yes, I used MeshLab. my question now is: why don't the visFile in xxxScene.json directly support ply or off format?

The simple answer is: because nobody had time to implement this yet ;-)

**Comment by Xayah-Hina** [comment]
@janbender I see it~ thank you guys for this awesome work~ ଘ(੭ˊᵕˋ)੭

**Comment by Xayah-Hina** [comment]
@korzen  I feel so sad that I still can not convert correctly using TetGen. T.T

in fact, I can successfully convert some simple model like `sphere.ply` and `dragon_vrip_res4.ply` using `tetgen -pq mymodel.ply` command. But when it comes to some complex model like `bun_zipper.ply` and my branch model `branch.ply`, it can't work properly any more...

terminal message:
<img width="647" alt="p1" src="https://user-images.githubusercontent.com/45065773/119214409-44deb080-baf9-11eb-9eca-61ecd771ec0b.png">

output files:
<img width="703" alt="p2" src="https://user-images.githubusercontent.com/45065773/119214428-5b850780-baf9-11eb-8b88-e57cdfac5f06.png">

my models:
[models.zip](https://github.com/InteractiveComputerGraphics/PositionBasedDynamics/files/6525787/models.zip)


I really wonder why ToT... Apologize to bother you but could you please offer me some advice for my confusion? Thanks a lot!

**Comment by janbender** [comment]
I'm no expert in using TetGen but maybe TetWild can give you better results: 

https://github.com/Yixin-Hu/TetWild
